### PR TITLE
recover

### PR DESCRIPTION
**Pull request by U1ltra**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,6 +6,7 @@ INCLUDE_DIR = include
 SRC_DIR = src
 BUILD_DIR = build
 TEST_DIR = test
+LOG_PERSIST = .persist
```
Comment (zhouzilong2020) [inline Makefile]: What is this for?

File: include/log_kv.h [hunk 2]
```diff
@@ -12,21 +12,26 @@
 
 #include "kv_store.h"
 
+const static char *RemovePrompt =
```
Comment (zhouzilong2020) [inline include/log_kv.h]: leftover

File: include/log_kv.h [hunk 3]
```diff
@@ -12,21 +12,26 @@
...
     /**
      * This function replaus the disk log to reconstruct the
      * key-value table, after a failure.
      */
     int recover();
 
+    /**
+     * failure simulates a machine failure by
+     */
+    void failure();
```
Comment (zhouzilong2020) [inline include/log_kv.h]: failure should be triggered outside the log_kv scope.

File: include/log_kv.h [hunk 4]
```diff
@@ -73,13 +81,58 @@ class LogKV : public KVStore
...
+            // if (entryCnt == 0)
+            //     printf("Recover Entry info %d %d %d | %s <-> %s\n",
+            //            logEntry->version, logEntry->keySize,
+            //            logEntry->valSize, (char *)&logEntry->payload, (char
+            //            *)&logEntry->payload + logEntry->keySize);
+
+            key = std::string((char *)&logEntry->payload);
+            val = std::string((char *)&logEntry->payload + logEntry->keySize);
+
+            if (logEntry->version == std::numeric_limits<uint16_t>::max())
```
Comment (zhouzilong2020) [inline include/log_kv.h:111]: Maybe make the special delete indicator a const?

File: src/log_kv.cpp [hunk 5]
```diff
@@ -65,7 +71,7 @@ void LogKV::deleteK(const std::string &key)
     }
 
     duplicatedEntryCnt++;
-    log->append(-1, key, NULL);
+    log->append(std::numeric_limits<uint16_t>::max(), key, NULL);
```
Comment (zhouzilong2020) [inline src/log_kv.cpp:74]: global const

File: src/log_kv.cpp [hunk 6]
```diff
@@ -76,3 +82,81 @@ size_t LogKV::size()
...
+                chunkCnt++;
+            }
+        }
+    }
+
+    log->recover(false);
+    return 0;
+}
+
+void LogKV::failure()
```
Comment (zhouzilong2020) [inline src/log_kv.cpp]: leftover

**Review by zhouzilong2020 — CHANGES_REQUESTED** [review]

**Review by zhouzilong2020 — APPROVED** [review]

**Comment by U1ltra** [comment]
king long!!!!

**Pull request merged**